src/java/org/apache/cassandra/db/guardrails/Guardrail.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.guardrails;
+
+import javax.annotation.Nullable;
+
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.service.ClientState;
+
+/**
+ * General class defining a given guardrail (hat guards against some particular usage/condition.
+ * <p>
+ * Some guardrails only emit warnings when triggered, while others fail the query that trigger them. Some may do one
+ * or the other based on specific threshold.
+ * <p>
+ * Note that all the defined classes support live updates, which is why each guardrail class constructor takes
+ * suppliers of the condition the guardrail acts on rather than the condition itself. This implies that said suppliers
+ * should be fast and non-blocking to avoid surprises.
+ */
+public interface Guardrail
+{
+    /**
+     * Checks whether this guardrail is enabled or not. This will be enabled if guardrails are
+     * ({@link Guardrails#enabled()}) and if the authenticated user (if specified) is not system nor superuser.
+     *
+     * @param state the query state, used to skip the check if the query is internal or is done by a superuser.
+     *              A {@code null} value means that the check should be done regardless of the query.
+     * @return {@code true} if this guardrail is enabled, {@code false} otherwise.
+     */
+    boolean enabled(@Nullable ClientState state);
```

Comment:
Nope, usages that are nor linked to a specific client query, such as compaction or repair, should simply pass a `null` `ClientState`. Additionally, these guardrails should probably use exclusively the soft limit, so we don't interrupt compaction/repair.

An example of this could be a guardrail for warning about large partitions detected during compaction ([CASSANDRA-17152](https://issues.apache.org/jira/browse/CASSANDRA-17152)). By the way, that guardrail could be a good candidate to use listeners if someone wants to get a more visible notification when that guardrail is triggered.

Reply:
> Nope, usages that are nor linked to a specific client query, such as compaction or repair, should simply pass a null ClientState

These usages are deeply nested, first example which comes to mind is RowIndexEntry (see org.apache.cassandra.db.RowIndexEntry.Serializer#checkSize).  The same method is used by compaction and the read path; we have 0 information about the caller.

> Additionally, these guardrails should probably use exclusively the soft limit, so we don't interrupt compaction/repair

@clohfink how do you feel about this statement?

Reply:
confirmed with Chris, just soft limits

Reply:
punting this problem for now, will need to work out when onboarding track_warnings and/or other ticket.